src/lib/log/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+//! This is a simple logger for oreboot, built on top of Rust embedded_hal.
+//! We expose the two macros `print!` and `println!` as well as the direct
+//! `_print` and `_debug` methods. The former is not inteded for use, while
+//! `_debug` serves as a fallback in case platform bringup gets really hard.
+//! Implement the embedded_hal non-blocking (nb) serial write trait, initialize
+//! your serial, and pass it along via `set_logger` to use the macros.
+//!
+//! ```rs
+//!     // MySerial implements embedded_hal::serial::nb::Write
+//!     let serial = init::MySerial::new(some_peripheral);
+//!     log::set_logger(serial);
+//!     log::_debug(42);
+//!     println!("oreboot 🦀");
+//! ```
+#![no_std]
+
+extern crate alloc;
+use alloc::boxed::Box;
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    fmt,
+    ptr::null_mut,
+};
+use embedded_hal::serial::{nb::Write, ErrorType};
+use nb::block;
+use spin::{Mutex, Once};
+
+pub trait Serial: ErrorType + Write {
+    /// This is meant to be the simplest fallback for debugging:
+    /// A "sign of life", possibly an LED flashing (GPIO high/low...),
+    /// an unforgiving write to a UART without polling for status or anything,
+    /// just to be sure that _something_ works. Can be a no-op, up to you.
+    fn debug(&self, num: u8);
+}
+
+/// Set the globally available logger that enables the macros.
+#[inline]
+pub fn set_logger<S: Serial<Error = Error> + 'static>(serial: S) {
+    unsafe {
+        LOGGER.0.call_once(|| LockedLogger {
+            l: Mutex::new(Box::new(serial)),
+        });
+    }
+}
+
+#[macro_export]
+macro_rules! print {
+    ($($arg:tt)*) => {
+        $crate::_print(core::format_args!($($arg)*));
+    }
+}
+
+#[macro_export]
+macro_rules! println {
+    () => ($crate::print!("\r\n"));
+    ($($arg:tt)*) => {
+        $crate::_print(core::format_args!($($arg)*));
+        $crate::print!("\r\n");
+    }
+}
+
+/// Error types that may happen with serial transfer, usually just `WouldBlock`
+#[derive(Debug)]
+pub struct Error {
+    pub kind: embedded_hal::serial::ErrorKind,
+}
+
+impl embedded_hal::serial::Error for Error {
+    #[inline]
+    fn kind(&self) -> embedded_hal::serial::ErrorKind {
+        self.kind
+    }
+}
+
+type SerialLogger = dyn Serial<Error = Error>;
+
+#[doc(hidden)]
+struct LockedLogger {
+    l: Mutex<Box<SerialLogger>>,
```

Comment:
`Pin` sounds interesting, I've not yet heard about it:
https://doc.rust-lang.org/std/pin/

I think I saw the `Box` thing in an example somewhere when trying to figure out how to implement this whole thing. :sweat_smile:

Reply:
Right, because without it, I would get
```
96 | static mut LOGGER: Logger = Wrap(Once::new());
   |                    ^^^^^^ doesn't have a size known at compile-time
```
and I couldn't figure out a way to work around it in hours; the simplest solution I found was `Box`.

I am trying to replace it with `Pin`, yielding
```
   --> src/lib/log/src/lib.rs:80:8
    |
80  |     l: Mutex<Pin<SerialLogger>>,
    |        ^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
```

An `Option` around a ref is what I had with the another implementation where the logger was passed the concrete UART type etc, and I couldn't get this here to work with it.

Reply:
You do need a dyn ref. See the code I sent you earlier? I'm also happy to chat about this shortyly.